sources/api/bootstrap-commands/src/main.rs

```diff
@@ -0,0 +1,315 @@
+
+Currently, Bottlerocket provides a set of actions (such as refresh-updates) that can be accessed
+through the apiserver, typically by running the apiclient in the control container. However, this
+new feature aims to extend the accessibility of these actions by allowing Bottlerocket consumers
+to run them prior to the bootstrap container execution.
+
+#Example:
+Given a bootstrap command called `001-test-bootstrap-commands` with the following configuration:
+
+[settings.bootstrap-commands.001-test-bootstrap-commands]
+commands = [[ "apiclient", "set", "motd=helloworld"]]
+essential = true
+mode = "always"
+
+This would update motd from "Welcome to Bottlerocket!" to "helloworld".
+*/
+
+#[macro_use]
+extern crate log;
+
+//use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use simplelog::{Config as LogConfig, LevelFilter, SimpleLogger};
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::BTreeMap;
+
+use std::env;
+use std::ffi::OsStr;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::{self, Command};
+use std::str::FromStr;
+
+use bottlerocket_modeled_types::{BootstrapMode, Identifier};
+use bottlerocket_settings_models::ValidCommand;
+
+const DEFAULT_CONFIG_PATH: &str = "/etc/bootstrap-commands/bootstrap-commands.toml";
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommandConfig {
+    bootstrap_commands: Option<BTreeMap<Identifier, BootstrapCommand>>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommand {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    commands: Option<Vec<ValidCommand>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    mode: Option<BootstrapMode>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    essential: Option<bool>,
+}
+
+/// Stores user-supplied global arguments
+struct Args {
+    log_level: LevelFilter,
+    config_path: PathBuf,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+            config_path: PathBuf::from_str(DEFAULT_CONFIG_PATH).unwrap(),
+        }
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn run_command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+
+    command
+        .args(args)
+        .status()
+        .context(error::ExecutionFailureSnafu { command })?;
+
+    Ok(())
+}
+
+/// Handles running of the api commands in a Bootstrap command
+fn handle_bootstrap_command<S>(name: S, bootstrap_command: &BootstrapCommand) -> Result<()>
+where
+    S: AsRef<str>,
+{
+    let name = name.as_ref();
+    let mode = bootstrap_command.mode.clone().unwrap_or_default();
+    let commands = bootstrap_command.commands.clone().unwrap_or_default();
```

Comment:
```suggestion
    let mode = bootstrap_command.mode.clone().unwrap_or_default();
    let commands = bootstrap_command.commands.as_deref().unwrap_or_default();
```

If you drop the `Option` from this struct and use `#[serde(default)]`, you could use `as_ref()` instead of `clone()` for the `mode` as well.